### ISSUE TITLE
Leak information in evaluation

### ISSUE THREAD

**Issue by ngavcc** [ztor2/lightgcn_recommender_pyg#1]
I see in your evaluation function, you use test edge to compute embedding of users and items. I think it is a bug because you can not use ground truth for testing. 
![Screenshot_3](https://github.com/ztor2/lightgcn_recommender_pyg/assets/129935403/6e4d2277-fde1-4961-837b-b73619dfccd3)
![Screenshot_2](https://github.com/ztor2/lightgcn_recommender_pyg/assets/129935403/d57d2e68-549f-49c7-88ac-8a6067c15df4)

**Comment by ztor2** [comment]
model.forward() in evaluation process do not change parameters which affect to inference.
It just returns aggregated embeddings of the test graph in the way that the paper suggests.
thank you.

**Comment by ngavcc** [comment]
I agree that i does not change parameters (which is embedding layer, or E0), but it does affect the value of A norm matrix. As a consequence, it changes the value of E (represent of user and item). You must not use ground truth of test set to compute user/item representation
![Uploading Screenshot_4.png…]()

**Comment by ngavcc** [comment]
And i see that you use embedding weight (which is E0) to present user/item, and compute score base on these representation. I think it should be E (weighted sum of E0, E1,..., En)
![Screenshot_5](https://github.com/ztor2/lightgcn_recommender_pyg/assets/129935403/8f5341aa-097f-4de7-a16d-88ef41e0292e)

**Comment by ngavcc** [comment]
![Screenshot_4](https://github.com/ztor2/lightgcn_recommender_pyg/assets/129935403/b5a74dc9-ad77-4c44-929c-90158afceb5e)
Formula in the paper

**Comment by 1197151063** [comment]
> And i see that you use embedding weight (which is E0) to present user/item, and compute score base on these representation. I think it should be E (weighted sum of E0, E1,..., En) ![Screenshot_5](https://private-user-images.githubusercontent.com/129935403/289407921-8f5341aa-097f-4de7-a16d-88ef41e0292e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI5NjgxNDQsIm5iZiI6MTcwMjk2Nzg0NCwicGF0aCI6Ii8xMjk5MzU0MDMvMjg5NDA3OTIxLThmNTM0MWFhLTA5N2YtNGRlNy1hMTZkLTg4ZWY0MWUwMjkyZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMjE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIxOVQwNjM3MjRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03Njc2ZTIxZDAxMzRmZjEzNjk2MmJkNWI5MjE3YjU4MGFiZmJlYTEwMGZjMzJhZDc1MWVlMzhmZjU2YzQ1MTIzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.GhsQUkN2N8WOsHQiUhcWnHCRwie5o5mlHmgMruCw8go)

I think the code here
```python
    user_embedding = model.user_emb.weight
    item_embedding = model.item_emb.weight
    rating = torch.matmul(user_embedding,item_embedding.T)
```
should be like 
```python
    user_embedding,_,item_embedding,_= model.forward(dataset.train_sparse_edge_index)
    rating = torch.matmul(user_embedding,item_embedding.T)
```

**Comment by ztor2** [comment]
@ngavcc @1197151063 
I checked this issue - user/item emb should come from E, not E0. 
This part will be fixed soon.

but I still didn't fully understand first issue(data leakage) suggested by @ngavcc.
I'll comment again after further review. thx.

**Comment by 1197151063** [comment]
> @ngavcc @1197151063 I checked this issue - user/item emb should come from E, not E0. This part will be fixed soon.
> 
> but I still didn't fully understand first issue(data leakage) suggested by @ngavcc. I'll comment again after further review. thx.

@ztor2 I think you are right in the evaluation part, I dont understand the leakage problem either. Haha

**Comment by ztor2** [comment]
I fixed the code which retrieving user/item embedding in evaluation process.
If I don't find any additional issues after further review, I'll close the issue.

**Comment by ztor2** [comment]
I fixed the code which retrieving user/item embedding in evaluation process.
If I don't find any additional issues after further review, I'll close the issue.

**Comment by ngavcc** [comment]
![Screenshot_6](https://github.com/ztor2/lightgcn_recommender_pyg/assets/129935403/84037cf1-40b2-4d29-8bf9-4f8fc5b3fe92)
As i see in your code, you use test_edges to compute user, item embedding; then compute scores based on this embedding. But i think that you must not use this information to predict. You can only use label in test set for compute metrics (like mrr, ndcg).
If you use this information; a simple "model" just output a high score (ex: 1000) for edges in test_edges, and low score (ex: -1000) for edges not in test_edges; can be a perfect model

**Comment by ngavcc** [comment]
suppose you bring your model to production. you need to use your model to predict items that your users will interacts in the future. all you have is interaction history of users in the past; and you can only use this history to compute embedding and scores. the validation process in your code likes you use user interactions in the future to compute embedding, scores. that make no sense

**Comment by ngavcc** [comment]
> > @ngavcc @1197151063 I checked this issue - user/item emb should come from E, not E0. This part will be fixed soon.
> > but I still didn't fully understand first issue(data leakage) suggested by @ngavcc. I'll comment again after further review. thx.
> 
> @ztor2 I think you are right in the evaluation part, I dont understand the leakage problem either. Haha

the author use test_sparse_edge_index to compute user/item embeddings, not use train_sparse_edge_index like you have suggested

**Comment by ngavcc** [comment]
![cap_3](https://github.com/ztor2/lightgcn_recommender_pyg/assets/129935403/b2cc34f4-7a5a-47b8-8624-500490340922)
![cap_2](https://github.com/ztor2/lightgcn_recommender_pyg/assets/129935403/5686caab-772b-4100-a9f5-7365a9114996)
![cap_1](https://github.com/ztor2/lightgcn_recommender_pyg/assets/129935403/c0098027-e441-46b4-a2b9-0a1a2dfd16aa)
@1197151063 can you recheck the author's code

**Comment by ztor2** [comment]
lol it is just a mistake in editing process.. I re-fixed and committed it.

I also found that the process in 'evaluation' function is almost unnecessary where @ngavcc first mentioned.
I removed this part - which could cause confusion and only left with 'get_metrics' function.
please let me know if there's any other problem.

**Comment by ztor2** [comment]
The code in this repository uses the test set for validation. 
If you want to compare the performance to the results of a paper strictly, the validation set should be extracted from the train set separately from the test set, and the test set should only be used for final evaluation. 

But when building a model for real-world use, after finishing initial evaluation, you don't always need to sample and use a test set(since user's behavior in the future can be used for evaluation). 

anyway I'll add a comment to prevent another confusion soon about this. thx.
